api/pkg/s3/datatype/acl.go

```diff
@@ -0,0 +1,248 @@
+				return acl, ErrUnsupportedAcl
+			}
+			canonUser = true
+		case ACL_TYPE_GROUP:
+			if grant.Grantee.URI == ACL_GROUP_TYPE_ALL_USERS {
+				log.Infoln("grant.Grantee.URI is", ACL_GROUP_TYPE_ALL_USERS)
+				if grant.Permission != ACL_PERM_READ {
+					log.Infoln("grant.Permission:", grant.Permission, "not equals", ACL_PERM_READ)
+					return acl, ErrUnsupportedAcl
+				}
+				acl = Acl{CannedAcl: ValidCannedAcl[CANNEDACL_PUBLIC_READ]}
+				group = true
+			} else if grant.Grantee.URI == ACL_GROUP_TYPE_AUTHENTICATED_USERS {
+				log.Infoln("grant.Grantee.URI is", ACL_GROUP_TYPE_AUTHENTICATED_USERS)
+				if grant.Permission != ACL_PERM_READ {
+					log.Infoln("grant.Permission:", grant.Permission, "not equals", ACL_PERM_FULL_CONTROL)
+					return acl, ErrUnsupportedAcl
+				}
+				acl = Acl{CannedAcl: ValidCannedAcl[CANNEDACL_AUTHENTICATED_READ]}
+				group = true
+			} else {
+				log.Infoln("grant.Grantee.URI is invalid:", grant.Grantee.URI)
+				return acl, ErrUnsupportedAcl
+			}
+		default:
+			log.Infoln("grant.Grantee.XsiType is invalid:", grant.Grantee.XsiType)
+			return acl, ErrUnsupportedAcl
+		}
+	}
+
+	if !canonUser {
+		log.Infoln("canonUser is invalid:", canonUser)
+		return acl, ErrUnsupportedAcl
+	}
+
+	if !group {
+		acl = Acl{CannedAcl: ValidCannedAcl[CANNEDACL_PRIVATE]}
+	}
+
+	return acl, nil
+}
+
+func createGrant(xsiType string, owner Owner, perm string, groupType string) (grant GrantResponse, err error) {
+
+	if xsiType == ACL_TYPE_CANON_USER {
+		grant.Grantee.ID = owner.ID
+		grant.Grantee.DisplayName = owner.DisplayName
+	} else if xsiType == ACL_TYPE_GROUP {
+		grant.Grantee.URI = groupType
+	} else {
+		return grant, ErrUnsupportedAcl
+	}
+	grant.Permission = perm
+	grant.Grantee.XmlnsXsi = XMLNSXSI
+	grant.Grantee.XsiType = xsiType
+	return
+}
+
+func CreatePolicyFromCanned(owner Owner, bucketOwner Owner, acl Acl) (
+	policy AccessControlPolicyResponse, err error) {
+
+	policy.ID = owner.ID
+	policy.DisplayName = owner.DisplayName
+	policy.Xmlns = XMLNS
+	grant, err := createGrant(ACL_TYPE_CANON_USER, owner, ACL_PERM_FULL_CONTROL, "")
+	if err != nil {
+		return policy, err
+	}
+	policy.AccessControlList = append(policy.AccessControlList, grant)
+	if acl.CannedAcl == "private" {
+		return policy, nil
+	}
+	switch acl.CannedAcl {
+	case "public-read":
+		owner := Owner{}
+		grant, err := createGrant(ACL_TYPE_GROUP, owner, ACL_PERM_READ, ACL_GROUP_TYPE_ALL_USERS)
+		if err != nil {
+			return policy, err
+		}
+		policy.AccessControlList = append(policy.AccessControlList, grant)
+	case "public-read-write":
+		owner := Owner{}
+		grant, err := createGrant(ACL_TYPE_GROUP, owner, ACL_PERM_READ, ACL_GROUP_TYPE_ALL_USERS)
+		if err != nil {
+			return policy, err
+		}
+		policy.AccessControlList = append(policy.AccessControlList, grant)
+		grant, err = createGrant(ACL_TYPE_GROUP, owner, ACL_PERM_WRITE, ACL_GROUP_TYPE_ALL_USERS)
+		if err != nil {
+			return policy, err
```

Comment:
if WRITE grant has failed, READ grant should be reverted

Reply:
@rhsakarpos  ok, I will fix it. thanks